aws/sdk/integration-tests/s3/tests/express.rs

```diff
@@ -98,3 +95,148 @@ async fn list_objects_v2_in_both_express_and_regular_buckets() {
+        .header("x-amz-create-session-mode", "ReadWrite")
+        .method("GET")
+        .body(SdkBody::empty())
+        .unwrap()
+}
+
+fn create_session_response() -> http::Response<SdkBody> {
+    http::Response::builder()
+        .status(200)
+        .body(SdkBody::from(
+            r#"<?xml version="1.0" encoding="UTF-8"?>
+            <CreateSessionResult xmlns="http://s3.amazonaws.com/doc/2006-03-01/">
+                <Credentials>
+                    <SessionToken>TESTSESSIONTOKEN</SessionToken>
+                    <SecretAccessKey>TESTSECRETKEY</SecretAccessKey>
+                    <AccessKeyId>ASIARTESTID</AccessKeyId>
+                    <Expiration>2024-01-29T18:53:01Z</Expiration>
+                </Credentials>
+            </CreateSessionResult>
+            "#,
+        ))
+        .unwrap()
+}
+
+fn operation_request_with_checksum(
+    query: &str,
+    kv: Option<(&str, &str)>,
+) -> http::Request<SdkBody> {
+    let mut b = http::Request::builder()
+        .uri(&format!("https://s3express-test-bucket--usw2-az1--x-s3.s3express-usw2-az1.us-west-2.amazonaws.com/{query}"))
+        .method("GET");
+    if let Some((key, value)) = kv {
+        b = b.header(key, value);
+    }
+    b.body(SdkBody::empty()).unwrap()
+}
+
+fn response_ok() -> http::Response<SdkBody> {
+    http::Response::builder()
+        .status(200)
+        .body(SdkBody::empty())
+        .unwrap()
+}
+
+#[tokio::test]
+async fn user_specified_checksum_should_be_respected() {
+    async fn runner(checksum: &str, value: &str) {
+        let http_client = StaticReplayClient::new(vec![
+            ReplayEvent::new(create_session_request(), create_session_response()),
+            ReplayEvent::new(
+                operation_request_with_checksum(
+                    "test?x-id=PutObject",
+                    Some((
+                        &format!("x-amz-checksum-{}", checksum.to_lowercase()),
+                        &format!("{value}"),
+                    )),
+                ),
+                response_ok(),
+            ),
+        ]);
+        let client = test_client(|b| b.http_client(http_client.clone())).await;
+
+        let _ = client
+            .put_object()
+            .bucket("s3express-test-bucket--usw2-az1--x-s3")
+            .key("test")
+            .body(SdkBody::empty().into())
+            .checksum_algorithm(ChecksumAlgorithm::from(checksum))
+            .send()
+            .await;
+
+        http_client.assert_requests_match(&[""]);
+    }
+
+    let crc32_checksum = "AAAAAA==";
+    let crc32c_checksum = "AAAAAA==";
+    let sha1_checksum = "2jmj7l5rSw0yVb/vlWAYkK/YBwk=";
+    let sha256_checksum = "47DEQpj8HBSa+/TImW+5JCeuQeRkm5NMpJWZG3hSuFU=";
+    for (checksum, value) in ChecksumAlgorithm::values().iter().zip(&[
+        crc32_checksum,
+        crc32c_checksum,
+        sha1_checksum,
+        sha256_checksum,
+    ]) {
```

Comment:
This seems like a fragile way to pair these values. Would be better to do something like:
```rust
let checksums = &[
    (ChecksumAlgorithm::Crc32, "AAAAAA=="),
    (ChecksumAlgorithm::Crc32c, "AAAAAA=="),
    // ...
];
assert_eq!(ChecksumAlgorithm::values().len(), checksums.len(), "checksum enum changed; need to update test");
for (checksum, value) in checksums {
    // ...
}
```
Although, that assert may be a bad idea since it could hold up a release upon adding a new algorithm? Not sure on that one since a new algorithm would require new code on our end anyway.

Reply:
Yeah, I prefer that. I would probably remove the assert as you suspected, since if the service adds a new checksum algorithm, it could fail in our release pipeline. It's safer for us to react to it and update the code rather than the change is pushed upon us during a release.